aten/src/ATen/native/cuda/Distributions.cu

```diff
@@ -0,0 +1,104 @@
+#include "ATen/NativeFunctions.h"
+#include "ATen/Dispatch.h"
+#include "ATen/cuda/CUDAApplyUtils.cuh"
+#include <curand.h>
+#include <curand_kernel.h>
+#include <curand_philox4x32_x.h>
+#include <thrust/functional.h>
+
+#include <THC/THCGeneral.h>
+#include <THC/THCHalf.h>
+#include <THC/THCApply.cuh>
+#include <THC/THCNumerics.cuh>
+#include <THC/THCTensorRandom.h>
+
+THCGenerator* THCRandom_getGenerator(THCState* state);
+
+namespace at {
+namespace native {
+
+namespace dist {
+  thrust::pair<uint64_t, uint64_t> get_philox_seed(Generator *gen) {
+    auto gen_ = THCRandom_getGenerator(at::globalContext().thc_state);
+    return thrust::make_pair(gen_->initial_seed, gen_->philox_seed_offset++);
+  }
+
+  // note that sample_poisson is adapted from Numpy's distributions.c
+  // see Distributions.cpp for the license
+  __device__ int64_t sample_poisson(float lambda, curandStatePhilox4_32_10_t *state) {
```

Comment:
I was thinking about this - the issue is the sampler required is different, but we could abstract it away using a function pointer (i.e. for getting a uniform, and getting a normal, etc.). The other thing is that the CPU version runs on double precision, but the GPU is at float, since it's slower to do double. Let me know what you think.

Reply:
Ok, I don't think I know how to do this. We need to pass a callable object to the shared `sample_poisson`, but that function needs to know `state` or have a reference to the `THGenerator`. So, it would need to be an `std::function`, which can't be called by `__device__` code. CUDA offers `nvstd::function`, but that only adds support for `__host__/__device__` shared calls in CUDA 8 (here's a link: https://devblogs.nvidia.com/parallelforall/new-compiler-features-cuda-8/)